src/BinSkim.Rules/PERules/BA6002.EliminateDuplicateStrings.cs

```diff
@@ -0,0 +1,147 @@
+            AnalysisApplicability result = AnalysisApplicability.NotApplicableToSpecifiedTarget;
+
+            reasonForNotAnalyzing = MetadataConditions.ImageIsILOnlyAssembly;
+            if (portableExecutable.IsILOnly) { return result; }
+
+            reasonForNotAnalyzing = MetadataConditions.ImageIsResourceOnlyBinary;
+            if (portableExecutable.IsResourceOnly) { return result; }
+
+            reasonForNotAnalyzing = MetadataConditions.CouldNotLoadPdb;
+            if (target.Pdb == null) { return result; }
+
+            reasonForNotAnalyzing = MetadataConditions.ImageIsNotBuiltWithMSVC;
+            if (!portableExecutable.IsTargetCompiledWithMSVC(target.Pdb)) { return result; }
+
+            reasonForNotAnalyzing = null;
+            return AnalysisApplicability.ApplicableToSpecifiedTarget;
+        }
+
+        public override void AnalyzePortableExecutableAndPdb(BinaryAnalyzerContext context)
+        {
+            PEBinary target = context.PEBinary();
+            Pdb pdb = target.Pdb;
+
+            var compilandsBinaryWithoutStringPooling = new List<ObjectModuleDetails>();
+            var compilandsLibraryWithoutStringPooling = new List<ObjectModuleDetails>();
+
+            foreach (DisposableEnumerableView<Symbol> omView in pdb.CreateObjectModuleIterator())
+            {
+                Symbol om = omView.Value;
+                ObjectModuleDetails omDetails = om.GetObjectModuleDetails();
+
+                if (omDetails.Language != Language.C &&
+                    omDetails.Language != Language.Cxx &&
+                    omDetails.Language != Language.MASM)
+                {
+                    continue;
+                }
+
+                if (!omDetails.HasDebugInfo)
+                {
+                    continue;
+                }
+
+                bool isMSVC = (omDetails.WellKnownCompiler == WellKnownCompilers.MicrosoftC ||
+                               omDetails.WellKnownCompiler == WellKnownCompilers.MicrosoftCxx);
+                if (isMSVC)
+                {
+                    if (!omDetails.EliminateDuplicateStringsEnabled)
+                    {
+                        CompilandRecord record = om.CreateCompilandRecord();
+                        if (!string.IsNullOrEmpty(record.Library))
+                        {
+                            compilandsLibraryWithoutStringPooling.Add(omDetails);
+                        }
+                        else
+                        {
+                            compilandsBinaryWithoutStringPooling.Add(omDetails);
+                        }
+                    }
+                }
+            }
+
+            if (compilandsLibraryWithoutStringPooling.Count > 0 || compilandsBinaryWithoutStringPooling.Count > 0)
+            {
+                if (compilandsLibraryWithoutStringPooling.Count > 0)
+                {
+                    GenerateCompilandsAndLog(context, compilandsLibraryWithoutStringPooling);
```

Comment:
>GenerateCompilandsAndLog [](http://example.com/codeflow?start=20&length=24)

There's nothing wrong with this pattern but I note that currently we can only configure a single severity for this check. A common pattern is for results that are directly relevant to a binary are 'errors' (because these are immediately fixable/under control of the user) while library dependencies remain as 'warnings' (because it is harder to solicit fixes for dependencies).

Note action here, just a thought to keep in mind...

Reply:
Yeah, I agree that the compiland list is a tricky one.  I tried doing the same thing for another rule (LTCG and the /GL option?) but the spew for things outside your control was too large so the signal/noise ratio was not good enough.